### PR TITLE
Convert to CsWin32 VARIANT

### PR DESCRIPTION
**Pull request by JeremyKuhne**
Converts VARIANT to CsWin32. I've left the files in the original location for now to keep the diff clear, will move them later.

It didn't become clear to me until I was fixing up tests that I could leverage using the newer C# feature of returning struct fields by ref to mitigate the nasty syntactical mess of replication of unions in C#. It's a nice pattern and one that might make sense moving into CsWin32.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/7977)

### CODE REVIEW

**Review by KlausLoeffelmann — APPROVED** [review]
Looks good to me so far, I just had one question.

File: src/System.Windows.Forms.Primitives/src/System.Windows.Forms.Primitives.csproj [hunk 1]
```diff
@@ -24,7 +24,7 @@
     <Deterministic>true</Deterministic>
     <ProduceReferenceAssembly>true</ProduceReferenceAssembly>
     <UsePublicApiAnalyzers>true</UsePublicApiAnalyzers>
-    <RootNamespace/>
+    <RootNamespace />
```
Comment (lonitra) [inline src/System.Windows.Forms.Primitives/src/System.Windows.Forms.Primitives.csproj:27]: spacing

File: src/System.Windows.Forms/src/System/Windows/Forms/AxHost.AxPerPropertyBrowsingEnum.cs [hunk 2]
```diff
@@ -98,9 +99,9 @@ private unsafe void EnsureArrays()
                                 continue;
                             }
 
-                            using var var = new Oleaut32.VARIANT();
+                            using VARIANT var = new();
                             HRESULT hr = ppb.GetPredefinedValue(_target.Dispid, cookie, &var);
-                            if (hr == HRESULT.S_OK && var.vt != Ole32.VARENUM.EMPTY)
+                            if (hr == HRESULT.S_OK && var.Type != VARENUM.VT_EMPTY)
```
Comment (lonitra) [inline src/System.Windows.Forms/src/System/Windows/Forms/AxHost.AxPerPropertyBrowsingEnum.cs:104]:
```suggestion
                            if (hr.Succeeded && var.Type != VARENUM.VT_EMPTY)
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ComponentModel/COM2Interop/Com2IPerPropertyBrowsingHandler.Com2IPerPropertyBrowsingEnum.cs [hunk 3]
```diff
@@ -97,9 +98,9 @@ private unsafe void EnsureArrays()
                                 continue;
                             }
 
-                            using var variant = new Oleaut32.VARIANT();
+                            using VARIANT variant = new();
                             HRESULT hr = ppb.GetPredefinedValue(_target.DISPID, cookie, &variant);
-                            if (hr == HRESULT.S_OK && variant.vt != Ole32.VARENUM.EMPTY)
+                            if (hr == HRESULT.S_OK && variant.Type != VARENUM.VT_EMPTY)
```
Comment (lonitra) [inline src/System.Windows.Forms/src/System/Windows/Forms/ComponentModel/COM2Interop/Com2IPerPropertyBrowsingHandler.Com2IPerPropertyBrowsingEnum.cs:103]:
```suggestion
                            if (hr.Succeeded && variant.Type != VARENUM.VT_EMPTY)
```

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.ActiveXImpl.cs [hunk 4]
```diff
@@ -1356,14 +1349,14 @@ internal void OnDocWindowActivate(BOOL fActivate)
             internal void OnFocus(bool focus)
             {
                 Debug.WriteLineIf(CompModSwitches.ActiveX.TraceInfo, "AXSource: SetFocus:  " + focus.ToString());
-                if (_activeXState[s_inPlaceActive] && _clientSite is IOleControlSite)
+                if (_activeXState[s_inPlaceActive] && _clientSite is Ole32.IOleControlSite)
                 {
-                    ((IOleControlSite)_clientSite).OnFocus(focus ? true : false);
+                    ((Ole32.IOleControlSite)_clientSite).OnFocus(focus ? true : false);
```
Comment (lonitra) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.ActiveXImpl.cs:1354]:
```suggestion
                    ((Ole32.IOleControlSite)_clientSite).OnFocus(focus);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/HtmlDocument.cs [hunk 5]
```diff
@@ -9,6 +9,7 @@
 using System.Drawing;
 using System.Globalization;
 using System.Runtime.InteropServices;
+using Windows.Win32.System.Com;
```
Comment (lonitra) [inline src/System.Windows.Forms/src/System/Windows/Forms/HtmlDocument.cs:12]: Would adding a global using result in too much clashing?

File: src/System.Windows.Forms.Primitives/src/Interop/OleAut32/Interop.VARIANT.cs [hunk 6]
```diff
@@ -1012,12 +905,12 @@ public struct VARIANTUnion
...
-
-            public struct VARIANTRecord
-            {
-                public void* pvRecord;
-                public IntPtr pRecInfo;
-            }
         }
     }
+
+    private struct VARIANTRecord
```
Comment (lonitra) [inline src/System.Windows.Forms.Primitives/src/Interop/OleAut32/Interop.VARIANT.cs:911]: I believe cswin32 has their own version of this struct. Is there a reason why we also want our own?

File: src/System.Windows.Forms/src/System/Windows/Forms/HtmlDocument.cs (hunk 5 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/HtmlDocument.cs:12]: Yes, and a fair amount of noise in areas that don't need it.

File: src/System.Windows.Forms.Primitives/src/Interop/OleAut32/Interop.VARIANT_BOOL.cs [hunk 7]
```diff
@@ -2,14 +2,15 @@
...
 
-internal partial class Interop
+namespace Windows.Win32.System.Com;
+
+// https://github.com/microsoft/win32metadata/issues/1319
+
+/// <summary>
+///  Wrapper for a VARIANT boolean type.
+/// </summary>
+internal enum VARIANT_BOOL : short
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Primitives/src/Interop/OleAut32/Interop.VARIANT_BOOL.cs:12]: We can yank it again when we update. Thanks for the heads up.

File: src/System.Windows.Forms.Primitives/src/Interop/OleAut32/Interop.VARIANTVector.cs [hunk 8]
```diff
@@ -4,43 +4,39 @@
 
 using System.Runtime.InteropServices;
 
-internal static partial class Interop
+namespace Windows.Win32.System.Com;
+
+internal unsafe ref struct VARIANTVector
 {
-    internal static partial class Oleaut32
+    public VARIANT[] Variants;
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Primitives/src/Interop/OleAut32/Interop.VARIANTVector.cs:11]: Yep

**Review by lonitra — APPROVED** [review]
🥳

**Pull request merged**